Don't show feedback snackbar till marketplace content has finished loading

andfinally opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes 17942-gh-Automattic/woocommerce.com.

- Ensures we only show the snackbar inviting feedback some time after the content has loaded.
  - We set a date `marketplace_redesign_2023_last_shown_date` in local storage to ensure we only show one snackbar.
- Ensures snackbar inherits fixed position from `.woocommerce-layout__footer`, so it's visible wherever you are on the page.
  - Removes `WooFooterItem` from around marketplace `Footer`, so it's no longer a child of the WooCommerce Admin footer.
  - Removes the `position: relative` from `.woocommerce-layout__footer`.
- Changes use of context. Context for marketplace now only has states for the selected tab and loading state.
  - Renames `ProductListContext`, `ProductListContextProvider` and `productListContextValue` to more generic `MarketplaceContext`, `MarketplaceContextProvider` and `marketplaceContextValue`.
  - We use this context in `Tabs` and `Content` to keep track of which tab is selected, and set the selected tab.
  - We also use it in `Discover` and `Extensions`, which both report loading state to the context. This allows us to only render the snackbar when loading is complete.
  - Extensions: moves `productList` and `setProductList` and logic for getting product list out of the context provider to this component. We don't need to share the list of products in the context.
- Renames hostname constant and creates constants for API paths.
 
<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. If you have a local storage value `marketplace_redesign_2023_last_shown_date` for your local WooCommerce environment, delete it. (It gets set whenever we show the snackbar, so delete it every time you want to test the snackbar again.)
2. Build WooCommerce Admin on your local environment e.g.: `cd plugins/woocommerce-admin` then `nvm use && pnpm start`.
3. Visit [WooCommerce > Marketplace > Extensions](http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions&tab=extensions).
4. Note that the snackbar with text "How easy is it to find an extension?" only appears about 7 seconds after the content loads on this screen.
5. Click on the `Browse` tab to go to the extensions screen.
6. Note that content for this screen loads, and the snackbar persists, and no second snackbar appears.
7. Delete the local storage value again and reload the page.
8. Note that the snackbar appears about 7 seconds after the content has loaded.
9. Note that the snackbar is positioned fixed relative to the viewport, so it's visible without scrolling.
10. Click around the tabs of the marketplace a bit more to satisfy yourself they're still working OK. Check that the footer looks like it does in trunk.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

Fixed rendering of snackbar inviting feedback on new style marketplace.

</details>

### Screenshots

<img width="500" src="https://github.com/woocommerce/woocommerce/assets/1647564/1bc4c04a-0fb6-4651-b614-d078b9c519d2">

andfinally committed:
[wccom-17942] Only showing feedback snackbar when content of in-app marketplace has finished loading. Making sure snackbar is fixed position, so it's visible wherever you are on the page.

- `ProductListContextProvider` provides `setIsLoading` function as well as `isLoading`.
- `Discover` uses these values from context, instead of keeping a loading state in itself.
- `FeedbackModal` calls `maybSetTimeout` when `isLoading` changes. If `isLoading` isn't truthy, and snackbar hasn't already rendered, it sets a timeout of 5 seconds to show it.

- Removed wrapping <WooFooterItem> from around Footer component, so it's no longer a child of the WooCommerce Admin `.woocommerce-layout__footer` footer.
- Removed the `position: relative` from `.woocommerce-layout__footer`. It needs to be `position: fixed`.
- Added FooterContent component to Footer, to allow the layout we want.

- Changed use of context. This now only has states for the selected tab and loading state.
- We use this context in `Tabs` and `Content` to keep track of which tab is selected, and set the selected tab.
- We also use it in `Discover` and `Extensions`, which both report loading state to the context. This allows us to use it to only render the snackbar when loading is complete.
- Extensions: moved `productList` and `setProductList` and logic for getting product list from the context provider to a state in this component. We don't need to share the list of products in the context.
- Renamed `ProductListContext`, `ProductListContextProvider` and `productListContextValue` to more generic `MarketplaceContext`, `MarketplaceContextProvider` and `marketplaceContextValue`.
- Renamed a constant and created constants for API paths.
- Only shows snackbar after content has loaded, and after a timeout. We set a date `marketplace_redesign_2023_last_shown_date` in local storage to ensure we only show one snackbar.

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/extensions/extensions.tsx:
```diff
@@ -1,23 +1,87 @@
+		const wccomSearchEndpoint =
+			MARKETPLACE_API_HOST +
+			MARKETPLACE_SEARCH_API_PATH + '?' +
+			params.toString();
+
+		// Fetch data from WCCOM API
+		fetch( wccomSearchEndpoint )
+			.then( ( response ) => response.json() )
+			.then( ( response ) => {
+				/**
+				 * Product card component expects a Product type.
+				 * So we build that object from the API response.
+				 */
+				const products = response.products.map(
+					( product: SearchAPIProductType ): Product => {
+						return {
+							id: product.id,
+							title: product.title,
+							description: product.excerpt,
+							vendorName: product.vendor_name,
+							vendorUrl: product.vendor_url,
+							icon: product.icon,
+							url: product.link,
+							// Due to backwards compatibility, raw_price is from search API, price is from featured API
+							price: product.raw_price ?? product.price,
+							averageRating: product.rating ?? 0,
+							reviewsCount: product.reviews_count ?? 0,
+						};
+					}
+				);
+
+				setProductList( products );
+			} )
+			.catch( () => {
+				setProductList( [] );
+			} )
+			.finally( () => {
+				setIsLoading( false );
+			} );
+	}, [ query ] );
```
This logic is moved here from `ProductListContextProvider` (now `MarketplaceContextProvider`). We don't need to keep this component's list of products in context. This matches more closely what we're doing in `Discover`.

raicem replied:
Oh nice! The less stuff in context, the better for now IMHO.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

kdevnel reviewed (approved):
This is a really nice refactor! I see now why it was a little bit of a wrestle. Thanks for tidying up the context and keeping Discover and Extensions similar - this sets us up well for Phase 2 with themes.

raicem commented on plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.tsx:
```diff
@@ -28,7 +28,7 @@ export default function HeaderAccount(): JSX.Element {
 	// Due to early erroring of TypeScript, it only takes the button version into account which doesn't accept href.
 	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	const accountURL: any = MARKETPLACE_URL + '/my-dashboard/';
+	const accountURL: any = MARKETPLACE_API_HOST + '/my-dashboard/';
```
Nitpick: maybe we should call this variable `MARKETPLACE_HOST`. 

`MARKETPLACE_API_HOST + '/my-dashboard/'` looked a little odd to me, considering we're creating a basic webpage URL and not an API endpoint.

andfinally replied:
Good point! Changed in [`6ae83b0` (#39869)](https://github.com/woocommerce/woocommerce/pull/39869/commits/6ae83b08ecb0d4f500996e50136c509ae1cd94f9).

raicem reviewed (approved):
These are great changes, done masterfully. Thanks @andfinally.

Dan-Q reviewed (approved):
Seconding @raicem's suggestion; otherwise this looks awesome.

andfinally committed:
[wccom-17942] Making sure snackbar doesn't appear before content has loaded.

- Addressing feedback. Renamed `MARKETPLACE_API_HOST` TO `MARKETPLACE_HOST` to make it less confusing – we don't only use it for API URLs.

andfinally committed:
[wccom-17942] Making sure snackbar doesn't appear before content has loaded.

- Fixing `prettier` linter errors.

andfinally committed:
[wccom-17942] Making sure snackbar doesn't appear before content has loaded.

- Fixing `prettier` linter errors.
